examples/fabric-admin/device_manager/UidGetter.cpp

```diff
@@ -0,0 +1,154 @@
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ */
+
+#include "UidGetter.h"
+
+using namespace ::chip;
+using namespace ::chip::app;
+using chip::app::ReadClient;
+
+namespace {
+
+void OnDeviceConnectedWrapper(void * context, Messaging::ExchangeManager & exchangeMgr, const SessionHandle & sessionHandle)
+{
+    reinterpret_cast<UidGetter *>(context)->OnDeviceConnected(exchangeMgr, sessionHandle);
+}
+
+void OnDeviceConnectionFailureWrapper(void * context, const ScopedNodeId & peerId, CHIP_ERROR error)
+{
+    reinterpret_cast<UidGetter *>(context)->OnDeviceConnectionFailure(peerId, error);
+}
+
+bool SuccessOrLog(CHIP_ERROR err, const char * name)
+{
+    if (err == CHIP_NO_ERROR)
+    {
+        return true;
+    }
+
+    ChipLogError(NotSpecified, "Failed to read %s: %" CHIP_ERROR_FORMAT, name, err.Format());
+
+    return false;
+}
+
+} // namespace
+
+UidGetter::UidGetter() :
+    mOnDeviceConnectedCallback(OnDeviceConnectedWrapper, this),
+    mOnDeviceConnectionFailureCallback(OnDeviceConnectionFailureWrapper, this)
+{}
+
+CHIP_ERROR UidGetter::GetUid(OnDoneCallback onDoneCallback, chip::Controller::DeviceController & controller, chip::NodeId nodeId,
+                             chip::EndpointId endpointId)
+{
+    assertChipStackLockedByCurrentThread();
+    VerifyOrDie(!mCurrentlyGettingUid);
+
+    mEndpointId       = endpointId;
+    mOnDoneCallback   = onDoneCallback;
+    mUniqueIdHasValue = false;
+    memset(mUniqueId, 0, sizeof(mUniqueId));
+    mCurrentlyGettingUid = true;
+
+    CHIP_ERROR err = controller.GetConnectedDevice(nodeId, &mOnDeviceConnectedCallback, &mOnDeviceConnectionFailureCallback);
+    if (err != CHIP_NO_ERROR)
+    {
+        mCurrentlyGettingUid = false;
+    }
+    return err;
+}
+
+void UidGetter::OnAttributeData(const ConcreteDataAttributePath & path, TLV::TLVReader * data, const StatusIB & status)
+{
+    VerifyOrDie(path.mClusterId == Clusters::BridgedDeviceBasicInformation::Id);
+
+    if (!status.IsSuccess())
+    {
+        ChipLogError(NotSpecified, "Response Failure: %" CHIP_ERROR_FORMAT, status.ToChipError().Format());
+        return;
+    }
+
+    switch (path.mAttributeId)
+    {
+    case Clusters::BridgedDeviceBasicInformation::Attributes::UniqueID::Id: {
+        mUniqueIdHasValue = SuccessOrLog(data->GetString(mUniqueId, sizeof(mUniqueId)), "UniqueId");
+        break;
+    }
+    default:
+        break;
+    }
+}
+
+void UidGetter::OnReportEnd()
+{
+    // We will call mOnDoneCallback in OnDone.
+}
+
+void UidGetter::OnError(CHIP_ERROR error)
+{
+    ChipLogProgress(NotSpecified, "Error Getting UID: %" CHIP_ERROR_FORMAT, error.Format());
+}
+
+void UidGetter::OnDone(ReadClient * apReadClient)
+{
+    mCurrentlyGettingUid = false;
+    std::optional<CharSpan> returnValue;
+    if (mUniqueIdHasValue)
+    {
+        returnValue = CharSpan(mUniqueId);
+    }
+    mOnDoneCallback(returnValue);
```

Comment:
nit:

```suggestion
    mOnDoneCallback(mUniqueIdHasValue ? std::make_optional<CharSpan>(mUniqueId) : std::nullopt);
```

may be more concise (if it compiles).